dist_run.py

```diff
@@ -235,31 +337,179 @@ def main():
+                dist.recv(x_recv, src, group=pp_group)
+                logger.info(f"RECEIVED {x_recv=}")
+                assert x_recv != 128006, f"next_token is header id={x_recv}"
+                _update_padded_sequence(padded_sequence, x_recv, res, prompt_lengths)
+                logger.info(
+                    f"Updated padded seq start: {prompt_lengths=}, {padded_sequence[:, prompt_lengths[0]-1]=}"
+                )
+
+            elif pp_rank == last_pp_group:
+                output = schedule.step()
+                decode_results = _batch_decode_next_tokens(
+                    output, prompt_lengths, tokenizer
+                )
+                logger.info(
+                    f"\n\n{color.green} {'Prefill' if step == 0 else '* Decode *'} responses ====>>>> {color.blue} {decode_results=} \n{color.reset}"
+                )
+
+                next_token = torch.tensor([decode_results[0][0]], device=device)
+                res.append(decode_results[0][1])
+
+                # increment prompt lengths for next token
+                for i in range(len(prompt_lengths)):
+                    prompt_lengths[i] += 1
+                    logger.info(
+                        f"output review {prompt_lengths[i]=}, {padded_sequence[i, prompt_lengths[i]-1]=}"
+                    )
+
+                if step < num_tokens - 1:
+                    dist.send(next_token, dst, pp_group)
+                logger.info(
+                    f"SENDING back...from rank={pp_rank} to dst={dst}, data {next_token.shape=}, {next_token=}"
+                )
+                assert next_token != 128006, f"next_token is header id={next_token}"
+
+            else:  # middle pp ranks
+                schedule.step()
+    """
+    with torch.no_grad():
+        for step in range(num_tokens):
+            # first
+            if pp_rank == 0:
+                schedule.step(padded_sequence)
+                # only receive if not last step
+                if step < num_tokens - 1:
+
+                    dist.recv(
+                        x_recv,
+                        src,
+                        group=pp_group,
+                    )
+                    _update_padded_sequence(
+                        padded_sequence, x_recv, res, prompt_lengths
+                    )
+
+            # last
+            elif pp_rank == last_pp_group:
+                output = schedule.step()
+                # need to decode the output
+                decode_results = _batch_decode_next_tokens(
+                    output=output, prompt_lengths=prompt_lengths, tokenizer=tokenizer
+                )
+                logger.info(
+                    f"\n\n{color.green} {'Prefill' if step == 0 else '* Decode *'} responses ====>>>> {color.blue} {decode_results=} \n{color.reset}"
+                )
+
+                next_token = torch.tensor([decode_results[0][0]], device=device)
+                res.append(decode_results[0][1])
```

Comment:
why `next_token` is from [0][0] while `res` is from [0][1]?

Reply:
this is b/c decode results returns both token_id (int) and token_str (readable).  
This was so we could push the text to the console during processing (and for me, debugging).